tcl/cast.tcl

```diff
@@ -853,4 +853,27 @@ namespace eval qc::cast {
         }
 	return -code error -errorcode CAST "Could not cast $string to an interval"
     }
+
+    proc next_url {string} {
+        if { ![regexp {^https?://} $string] } {
+            # Relative url            
+            set next_url [string trimleft $string /]
+            set next_url "[qc::conn_location]/$next_url"           
+        } else {
+            set next_url $string
+        }
+
+        # check for malicious mal-formed url
+        if { ![qc::is url $next_url] } {
+            return -code error -errorcode CAST "Could not cast \"[html_escape $next_url]\" to valid next_url. Invalid URL."
+        }
+        
+        # check that redirection is to the same domain
+        set conn_host [qc::conn_host]
+        if { ![regexp "^https?://${conn_host}(:\[0-9\]+)?(\\?|#|/|\$)" $next_url] } {
+            return -code error -errorcode CAST "Could not cast \"[html_escape $next_url]\" to valid next_url. Domain must be \"$conn_host\"."
```

Comment:
$conn_host can be controlled by header manipulation so is not completely safe - does this end up being rendered back to the user? Is so, it need escaping (Maybe something further up takes care of that?)

Reply:
Yes this end up being rendered back to the user and yes I think it should be the caller's responsibility to take care of that. Would it be okay to escape the next_url before it is passed to this cast api [here](https://github.com/qcode-software/qcode-tcl/pull/715/files#diff-bb1f05244dfece6f46f7e54ddff8d5a419586b13eca9f5294fc8a3229884a27fR135)

Reply:
We're pulling ```conn_host``` directly out of the headers ```[qc::conn_host]```within this proc so the caller can't deal with it.

Reply:
Good point. I can `html_escape` the `string` at the start of the API if that helps ?

Reply:
If we cannot trust the `conn_host` the only other option I can think of is using a `param`.